core/src/main/java/ai/timefold/solver/core/impl/neighborhood/stream/enumerating/bi/JoinBiEnumeratingStream.java

```diff
@@ -45,26 +45,13 @@ public void collectActiveEnumeratingStreams(Set<AbstractEnumeratingStream<Soluti
     public void buildNode(DataNodeBuildHelper<Solution_> buildHelper) {
         var solutionView = buildHelper.getSessionContext().solutionView();
         var filteringDataJoiner = this.filtering == null ? null : this.filtering.toBiPredicate(solutionView);
-        var outputStoreSize = buildHelper.extractTupleStoreSize(this);
         TupleLifecycle<BiTuple<A, B>> downstream = buildHelper.getAggregatedTupleLifecycle(childStreamList);
         var indexerFactory = new IndexerFactory<>(joiner.toBiJoiner());
+        var positionTracker =
+                buildHelper.getTupleStorePositionTracker(this, leftParent.getTupleSource(), rightParent.getTupleSource());
         var node = indexerFactory.hasJoiners()
-                ? new IndexedJoinBiNode<>(indexerFactory,
-                        buildHelper.reserveTupleStoreIndex(leftParent.getTupleSource()),
-                        buildHelper.reserveTupleStoreIndex(leftParent.getTupleSource()),
-                        buildHelper.reserveTupleStoreIndex(leftParent.getTupleSource()),
-                        buildHelper.reserveTupleStoreIndex(rightParent.getTupleSource()),
-                        buildHelper.reserveTupleStoreIndex(rightParent.getTupleSource()),
-                        buildHelper.reserveTupleStoreIndex(rightParent.getTupleSource()),
```

Comment:
Through the refactoring, I was aiming at this craziness, which was getting propagated through 3 layers of inheritance. Now each node is responsible for reserving their own indexes.